### PR TITLE
updates to tls/uij magnitude optimisation

### PR DESCRIPTION
**Pull request by nicholas-pearce**
-- added lasso and ridge regression penalties
-- added/expanded tests for new functions
-- fix white-space/indentation to be consistent 2-space throughout
-- rename variable names for clarity

all tests run on linux cleanly
libtbx.find_clutter runs cleanly
libtbx.find_unused_imports_crude runs cleanly
should be python3-compatible but not tested

### CODE REVIEW

**Comment by bkpoon** [comment]
I took a quick look. There is one Python 3 syntax change ("iteritems" to "items").

https://github.com/cctbx/cctbx_project/blob/de1a55ecfd34182cec36dc9a9cd5ef5d71e4885d/mmtbx/tls/optimise_amplitudes.py#L52

I am able to run the test in Python 3.7 after that change.

 I'm currently at SACLA, but I can take a closer look near the end of next week. Maybe @olegsobolev can take a look as well.

One comment is that can the `OptimisationWeights` be a subclass of `dict`? Or have a parent class that has the same properties of `dict`? Then you won't need to have separate functions for `generic` types and `dict` types.

File: mmtbx/tls/optimise_amplitudes.py [hunk 1]
```diff
@@ -11,77 +11,208 @@
...
+    Transfer weights from an input object (dict or other OptimisationWeights object)
+    """
+    if weights is None:
+      return self
+    elif isinstance(weights, dict):
+      return self.update_from_dict(weights)
+    elif isinstance(weights, type(self)):
+      return self.update_from_dict(weights.__dict__)
+    else:
+      raise Exception('Unknown type for `weights`')
```
Comment (Anthchirp) [inline mmtbx/tls/optimise_amplitudes.py]: Should be `ValueError`

File: mmtbx/tls/optimise_amplitudes.py [hunk 2]
```diff
@@ -11,77 +11,208 @@
...
+    """
+    self = cls()
+    found_none = True
+    for k in self.__dict__.keys():
+      v = getattr(obj, k, None)
+      if v is None: continue
+      found_none = False
+      self.__dict__[k] = v
+    if found_none is True:
+      raise Exception('No overlapping attributes were found between objects')
```
Comment (Anthchirp) [inline mmtbx/tls/optimise_amplitudes.py]: `ValueError`

**Comment by olegsobolev** [comment]
Tests work in my hands. Looks good to me.

**Comment by nicholas-pearce** [comment]
OK — great. Thanks to all for checking and for feedback. 

I’ll change the “not-a-dict” class as suggested — a named tuple seems appropriate. That commit will come early next week, probably.

Thanks,
Nick

**Comment by bkpoon** [comment]
Thanks for making the changes!

**Pull request merged**